plugins/producer_plugin/include/eosio/producer_plugin/block_timing_util.hpp

```diff
@@ -1,13 +1,42 @@
 #pragma once
 #include <eosio/chain/block_timestamp.hpp>
 #include <eosio/chain/config.hpp>
+#include <eosio/chain/producer_schedule.hpp>
 
 namespace eosio {
 
 enum class pending_block_mode { producing, speculating };
 
 namespace block_timing_util {
 
+   // Store watermarks
+   // Watermarks are recorded times that the specified producer has produced.
+   // Used by calculate_producer_wake_up_time to skip over already produced blocks avoiding duplicate production.
+   class producer_watermarks {
+   public:
+      void consider_new_watermark(chain::account_name producer, uint32_t block_num, chain::block_timestamp_type timestamp) {
+         auto itr = _producer_watermarks.find(producer);
+         if (itr != _producer_watermarks.end()) {
+            itr->second.first  = std::max(itr->second.first, block_num);
+            itr->second.second = std::max(itr->second.second, timestamp);
+         } else {
+            _producer_watermarks.emplace(producer, std::make_pair(block_num, timestamp));
+         }
```

Comment:
```suggestion
         auto& w = _producer_watermarks[producer];
         w = std::max(w, std::make_pair(block_num, timestamp));
```

Reply:
```
Example:

block_slot = 13 - time 2023-08-11T10:00:00.000 - block_num 50
block_slot = 14 - time 2023-08-11T10:00:00.500 - block_num 51
block_slot = 15 - time 2023-08-11T10:00:01.000 - block_num 52
!!FORK!!
block_slot = 16 - time 2023-08-11T10:00:01.500 - block_num 50 <- lower block_num, but higher block_slot
block_slot = 17 - time 2023-08-11T10:00:02.000 - block_num 51
```

Reply:
Oh my I didn't think of that, you are right. Thanks for explaining.